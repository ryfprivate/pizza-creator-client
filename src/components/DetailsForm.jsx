import React from 'react';
import styled from 'styled-components';
import Input from './Input';

const DetailsFormDiv = styled.div`
  display: grid;
  grid-template-columns: repeat(3, 1fr);
  grid-gap: 1rem;
`;

export default class DetailsForm extends React.Component {
  render() {
    const { data, onDataChange, dirty } = this.props;

    const confirmMobileValidationMessage = getValidationMessageForConfirmMobile(data.confirmMobile, data.mobile);

    return (
      <DetailsFormDiv>
        <Input
          label="Name"
          value={data.name}
          formDirty={dirty}
          validate={data.name}
          onChange={value => onDataChange('name', value)}
          validationMessage="Please enter your name"
        />
        <Input
          label="Mobile Number"
          value={data.mobile}
          formDirty={dirty}
          validate={!getValidationMessageForMobile(data.mobile)}
          onChange={value => onDataChange('mobile', value)}
          validationMessage={getValidationMessageForMobile(data.mobile)}
        />
        <Input
          label="Confirm Mobile"
          value={data.confirmMobile}
          formDirty={dirty}
          validate={!confirmMobileValidationMessage}
          onChange={value => onDataChange('confirmMobile', value)}
          validationMessage={confirmMobileValidationMessage}
        />
        <Input
          label="Email"
          value={data.email}
          formDirty={dirty}
          validate={!getValidationMessageForEmail(data.email)}
          onChange={value => onDataChange('email', value)}
          validationMessage={getValidationMessageForEmail(data.email)}
        />
        <Input
          label="Address"
          value={data.address}
          formDirty={dirty}
          validate={data.address}
          onChange={value => onDataChange('address', value)}
          validationMessage="Please enter your Address"
        />
        <Input
          label="Post Code"
          value={data.postCode}
          formDirty={dirty}
          validate={!getValidationMessageForPostCode(data.postCode)}
          onChange={value => onDataChange('postCode', value)}
          validationMessage={getValidationMessageForPostCode(data.postCode)}
        />
      </DetailsFormDiv>
    );
  }
}

function validatePostCode(postCode) {
  var re = /^\d{4}$/;
  return re.test(String(postCode).toLowerCase());
}

function getValidationMessageForPostCode(postCode) {
  if (!postCode) {
    return 'Please enter your Post Code';
  }

  if (!validatePostCode(postCode)) {
    return 'The Post Code format is invalid';
  }

  return '';
}

function validateEmail(email) {
  var re = /^(([^<>()\[\]\\.,;:\s@"]+(\.[^<>()\[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/;
  return re.test(String(email).toLowerCase());
}

function getValidationMessageForEmail(email) {
  if (!email) {
    return 'Please enter your Email';
  }

  if (!validateEmail(email)) {
    return 'The Email format is invalid';
  }

  return '';
}

function validateMobile(mobile) {
  var re = /^04\d{8}$/;
  return re.test(String(mobile).toLowerCase());
}

function getValidationMessageForConfirmMobile(confirmMobile, mobile) {
  if (!confirmMobile) {
    return 'Please enter your Confirm Mobile';
  }

  if (confirmMobile !== mobile) {
    return 'The Confirm Mobile is not same as Mobile Number';
  }

  return '';
}

function getValidationMessageForMobile(mobile) {
  if (!mobile) {
    return 'Please enter your Mobile';
  }

  if (!validateMobile(mobile)) {
    return 'The Mobile format is invalid';
  }

  return '';
}
