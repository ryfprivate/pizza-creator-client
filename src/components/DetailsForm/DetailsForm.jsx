import React from 'react';
import styled from 'styled-components';
import Input from '../Input/Input';

const DetailsFormDiv = styled.div`
  display: grid;
  grid-template-columns: repeat(3, 1fr);
  grid-gap: 1rem;
`;

function validatePostCode(postCode) {
  var re = /^\d{4}$/;
  return re.test(String(postCode).toLowerCase());
}

function getValidationMessageForPostCode(postCode) {
  if (!postCode) {
    return 'Please enter your Post Code';
  }

  if (!validatePostCode(postCode)) {
    return 'The Post Code format is invalid';
  }

  return '';
}

function validateEmail(email) {
  var re = /^(([^<>()\[\]\\.,;:\s@"]+(\.[^<>()\[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/;
  return re.test(String(email).toLowerCase());
}

function getValidationMessageForEmail(email) {
  if (!email) {
    return 'Please enter your Email';
  }

  if (!validateEmail(email)) {
    return 'The Email format is invalid';
  }

  return '';
}

function getValidationMessageForConfirmEmail(confirmEmail, email) {
  if (!confirmEmail) {
    return 'Please confirm your Email';
  }

  if (confirmEmail !== email) {
    return 'The Confirm Email does not match';
  }

  return '';
}

function validateMobile(mobile) {
  var re = /^04\d{8}$/;
  return re.test(String(mobile).toLowerCase());
}

function getValidationMessageForMobile(mobile) {
  if (!mobile) {
    return 'Please enter your Mobile';
  }

  if (!validateMobile(mobile)) {
    return 'The Mobile format is invalid';
  }

  return '';
}

const DetailsForm = ({ data, dirty, onDataChange }) => {
  const confirmEmailValidationMessage = getValidationMessageForConfirmEmail(
    data.confirmEmail,
    data.email
  );

  return (
    <DetailsFormDiv>
      <Input
        label="Name"
        value={data.name}
        formDirty={dirty}
        validate={data.name}
        onChange={(value, showError) => onDataChange('name', value, showError)}
        validationMessage="Please enter your name"
      />
      <Input
        label="Email"
        value={data.email}
        formDirty={dirty}
        validate={!getValidationMessageForEmail(data.email)}
        onChange={(value, showError) => onDataChange('email', value, showError)}
        validationMessage={getValidationMessageForEmail(data.email)}
      />
      <Input
        label="Confirm Email"
        value={data.confirmEmail}
        formDirty={dirty}
        validate={!confirmEmailValidationMessage}
        onChange={(value, showError) =>
          onDataChange('confirmEmail', value, showError)
        }
        validationMessage={confirmEmailValidationMessage}
      />
      <Input
        label="Address"
        value={data.address}
        formDirty={dirty}
        validate={data.address}
        onChange={(value, showError) =>
          onDataChange('address', value, showError)
        }
        validationMessage="Please enter your Address"
      />
      <Input
        label="PostCode"
        value={data.postCode}
        formDirty={dirty}
        validate={!getValidationMessageForPostCode(data.postCode)}
        onChange={(value, showError) =>
          onDataChange('postCode', value, showError)
        }
        validationMessage={getValidationMessageForPostCode(data.postCode)}
      />
      <Input
        label="Mobile Number"
        value={data.mobile}
        formDirty={dirty}
        validate={!getValidationMessageForMobile(data.mobile)}
        onChange={(value, showError) =>
          onDataChange('mobile', value, showError)
        }
        validationMessage={getValidationMessageForMobile(data.mobile)}
      />
    </DetailsFormDiv>
  );
};

export default DetailsForm;
