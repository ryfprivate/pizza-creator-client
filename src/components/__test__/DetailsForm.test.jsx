import React from "react";
import { mount } from "enzyme";
import DetailsForm from "../DetailsForm";

const props = {
  data: {},
  dirty: false,
  onDataChange: jest.fn()
};

let wrapper;

describe("test DetailsForm component", () => {
  beforeEach(() => {
    wrapper = mount(<DetailsForm {...props} />);
  });

  it("should render the component to match snapshot", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("should render correct amount of inputs", () => {
    const inputList = wrapper.find("input");
    expect(inputList.length).toEqual(6);
  });

  it("should receive correct props", () => {
    expect(wrapper.props().dirty).toBe(false);
  });

  it("should trigger the correct function callback", () => {
    const firstInput = wrapper.find("input").first();
    firstInput.simulate("change");
    expect(props.onDataChange).toHaveBeenCalledWith("name", firstInput.props().value);
  });
});
