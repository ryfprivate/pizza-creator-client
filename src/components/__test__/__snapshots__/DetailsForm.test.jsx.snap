// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test DetailsForm component should render the component to match snapshot 1`] = `ReactWrapper {}`;
